index.js

```diff
@@ -115,24 +114,35 @@ function Client (opts) {
   this.config(opts)
   this._log = this._conf.logger || new NoopLogger()
 
-  // start stream in corked mode, uncork when cloud
-  // metadata is fetched and assigned.  Also, the
-  // _maybeUncork will not uncork until _encodedMetadata
-  // is set
-  this.cork()
-  this._fetchAndEncodeMetadata(() => {
-    // _fetchAndEncodeMetadata will have set/memoized the encoded
-    // metadata to the _encodedMetadata property.
-
-    // This reverses the cork() call in the constructor above. "Maybe" uncork,
-    // in case the client has been destroyed before this callback is called.
-    this._maybeUncork()
-
-    // the `cloud-metadata` event allows listeners to know when the
-    // agent has finished fetching and encoding its metadata for the
-    // first time
-    this.emit('cloud-metadata', this._encodedMetadata)
-  })
+  if (this._conf.cloudMetadataFetcher && this._conf.expectExtraMetadata) {
+    throw new Error('it is an error to create a Client with both cloudMetadataFetcher and expectExtraMetadata')
+  } else if (this._conf.cloudMetadataFetcher) {
+    // Start stream in corked mode, uncork when cloud metadata is fetched and
+    // assigned.  Also, the _maybeUncork will not uncork until _encodedMetadata
+    // is set.
+    this._log.trace('corking (cloudMetadataFetcher)')
+    this.cork()
+    this._fetchAndEncodeMetadata(() => {
+      // _fetchAndEncodeMetadata will have set/memoized the encoded
+      // metadata to the _encodedMetadata property.
+
+      // This reverses the cork() call in the constructor above. "Maybe" uncork,
+      // in case the client has been destroyed before this callback is called.
+      this._maybeUncork()
+      this._log.trace('uncorked (cloudMetadataFetcher)')
+
+      // the `cloud-metadata` event allows listeners to know when the
+      // agent has finished fetching and encoding its metadata for the
+      // first time
+      this.emit('cloud-metadata', this._encodedMetadata)
+    })
+  } else if (this._conf.expectExtraMetadata) {
+    // Uncorking will happen in the expected `.setExtraMetadata()` call.
+    this._log.trace('corking (expectExtraMetadata)')
+    this.cork()
+  } else {
+    this._resetEncodedMetadata()
```

Comment:
This call to `this._resetEncodedMetadata` wasn't present in the previous version of the object constructor and I don't completely follow what it's doing here now.  No need for a change, but I would like to understand what use case it handles or bug it fixes.

Reply:
It handles the case where the Agent uses the client but passes neither `expectExtraMetadata`, nor `cloudMetadataFetcher`. This case is the empty "else"-block here: https://github.com/elastic/apm-agent-nodejs/pull/2407/files#diff-d5109920b34c2b2e141bd41232d8876432a17479c388e142d8a036bb7b36ea62R792-R800

Otherwise `this._encodedMetadata` is never set and the first intake request would fail.  For the other two cases `_resetEncodedMetadata` is called (a) when the cloudMetadataFetcher calls back or (b) when `client.setExtraMetadata()` is called.